consensus/src/block_storage/sync_manager.rs

```diff
@@ -721,22 +812,51 @@ impl BlockRetriever {
                     result_blocks.extend(batch);
                     break;
                 },
-                _e => {
+                res => {
                     bail!(
-                        "Failed to fetch block {}, for original start {}",
+                        "Failed to fetch block {}, for original start {}, returned status {:?}",
                         last_block_id,
                         block_id,
+                        res
                     );
                 },
             }
         }
+
+        // Confirm retrieval hit the first block we care about
         assert_eq!(
-            result_blocks
-                .last()
-                .expect("Expected at least a result_block")
-                .id(),
-            target_block_id
+            result_blocks.first().expect("blocks are empty").id(),
+            block_id,
+            "Expecting in the retrieval response, first block should be {}, but got {}",
+            block_id,
+            result_blocks.first().expect("blocks are empty").id(),
         );
+
+        // Confirm retrieval hit the last block/round we care about
+        // Slightly different logic if using execution pool and not
+        match target_block_retrieval_payload {
+            TargetBlockRetrieval::TargetBlockId(target_block_id) => {
+                assert_eq!(
+                    result_blocks
+                        .last()
+                        .expect("Expected at least a result_block")
+                        .id(),
+                    target_block_id
+                );
+            },
+            TargetBlockRetrieval::TargetRound(target_round) => {
+                let last_block = result_blocks.last().expect("blocks are empty");
+                assert!(
```

Comment:
also i think assert is probably too strong here, in case we don't handle malicious response correctly